Add multiple process items page for Android (#154)
smartdevicelink/sdl_app_library_guides: docs/Frequently Asked Questions/Multiple Process Items/index.md

# Multiple Process Items

The SmartDeviceLink Android library uses multiple process and there are some items that should be understood about why that is necessary and precautions to take while handling that situation. 

## Why does the router service run in its own process?

> The router service is designed to live outside the normal lifecycle of the app integrating the SDL framework. The different process allows a level of security to cut off access to the hosting application's data. It also allows the router service to not interfere with the hosting applications runtime; this means if the router service unexpectedly stops or crashes, it will not take down the entire app. This relationship also works in the opposite direction, which is important to maintain a good user experience when apps are connected through a router service.

Reviewer: ```suggestion
The router service is designed to live outside the normal lifecycle of the app integrating the SDL framework. The different process allows a level of security to cut off access to the hosting application's data. It also allows the router service to not interfere with the hosting application's runtime; this means if the router service unexpectedly stops or crashes, it will not take down the entire app. This relationship also works in the opposite direction, which is important to maintain a good user experience when apps are connected through a router service.
```

> ## Content providers and multi-process issues

Reviewer: Is "multi-process" the same as "multiple process" referenced earlier?  If so, I'd make sure this reference is consistent throughout the document.

> ## Content providers and multi-process issues

Author: ```suggestion
## Content providers and multiple process issues
```

> Android content providers have a unique lifecycle that does not work in the expected flow. Content providers are actually started before the Application class and following Activities, Services, etc. Some libraries use this to know when their code/module can initialize and always be ready for the entire lifecycle of the application. This is found with many of Google libraries (Firebase, Jetpack, etc).

Reviewer: ```suggestion
Android content providers have a unique lifecycle that does not work in the expected flow. Content providers are actually started before the `Application` class and following Activities, Services, etc. Some libraries use this to know when their code/module can initialize and always be ready for the entire lifecycle of the application. This is found with many Google libraries (Firebase, Jetpack, etc).
```

The issue is that, by default, content providers are only attached to the main process. This process is the same as the application package unless otherwise specified. This means, when the main process starts the content provider will be started, but if a different process is started the content provider will not be started. So if the app has its first start from a component that is designed to run in a different process, the content provider won't be ready by the time those components start up; this includes the application instance for that process. 

### Why is this a problem?

> The issue is when there is code in a developer's custom `Application` class that assumes the content provider has already been initialized but an instance of that child `Application` class is created for a process outside the main process.

Reviewer: ```suggestion
The issue occurs when there is code in a developer's custom `Application` class that assumes the content provider has already been initialized, but an instance of that child `Application` class is created for a process outside of the main process.
```

For example:

```
public class MyApplication extends Application {

    @Override
    public void onCreate() {
        super.onCreate();
        
         ModuleUsingContentProviderForInit.doSomething();

    }
}

```

If an instance is created outside the main process, this application will crash with a runtime exception. 

### Workaround

> Depending on the module that uses a content provider for initialization it could be possible to start/initialize it from the `onCreate` method of the extended `Application` class. It should be noted that the module would then need to be setup for a multiple process environment, which is not always the case.

Reviewer: ```suggestion
Depending on the module that uses a content provider for initialization, it could be possible to start/initialize it from the `onCreate` method of the extended `Application` class. It should be noted that the module would then need to be set up for a multiple process environment, which is not always the case.
```

        	return;
        }
        
        ModuleUsingContentProviderForInit.doSomething();

    }
    
    /**
    *
    * @return if this process is the SDL router service process
    */
    private boolean isSdlProcess(){
        int myPid = android.os.Process.myPid();
        ActivityManager am = (ActivityManager)this.getSystemService(ACTIVITY_SERVICE);
        if (am == null || am.getRunningAppProcesses() == null) {
            return false;
        }
        
        for (ActivityManager.RunningAppProcessInfo processInfo : am.getRunningAppProcesses()) {
            if (processInfo != null && processInfo.pid == myPid) {
                return ROUTER_SERVICE_PROCESS.equals(processInfo.processName);
            }
        }
        return false;
    }
}

```

!!! NOTE
If other callback methods in your `Application` class are used, they must also check this flag to prevent unintended behavior.
!!!

The use of this flag can help prevent errors when extending the `Application` class that assume it always has the main process started first. This solution could be modified to change the flag to monitor if this is the main process or not very easily. 

> ### Custom `Application` classes instance for each process

Reviewer: The code block in a header formats a little funny.  You may want to consider removing the ticks.

![Screen Shot 2019-10-03 at 10 54 32 AM](https://user-images.githubusercontent.com/22600496/66138217-3bbf6980-e5cc-11e9-961d-258937e6c69a.png)

> ### Custom `Application` classes instance for each process

Author: ```suggestion
### Custom Application classes instance for each process
```
